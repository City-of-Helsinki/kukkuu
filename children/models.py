from hashlib import sha256
from typing import Iterable, Optional

from django.conf import settings
from django.core.validators import RegexValidator
from django.db import models, transaction
from django.db.models import Q
from django.utils import timezone
from django.utils.translation import gettext_lazy as _

from common.models import TimestampedModel, UUIDPrimaryKeyModel
from languages.models import Language
from users.models import Guardian


class ChildQuerySet(models.QuerySet):
    def user_can_view(self, user):
        return self.filter(
            Q(guardians__user=user) | Q(project__in=user.administered_projects)
        ).distinct()

    def user_can_update(self, user):
        return self.filter(guardians__user=user)

    def user_can_delete(self, user):
        return self.filter(guardians__user=user)

    @transaction.atomic()
    def delete(self):
        for child in self:
            child.delete()


postal_code_validator = RegexValidator(
    regex=r"^\d{5}$", message="Postal code must be 5 digits", code="invalid_postal_code"
)


class Child(UUIDPrimaryKeyModel, TimestampedModel):
    name = models.CharField(verbose_name=_("name"), max_length=64, blank=True)
    birthyear = models.DateField(verbose_name=_("birthyear"))
    postal_code = models.CharField(
        verbose_name=_("postal code"),
        max_length=5,
        validators=[postal_code_validator],
    )
    guardians = models.ManyToManyField(
        Guardian,
        verbose_name=_("guardians"),
        related_name="children",
        through="children.Relationship",
        blank=True,
    )
    project = models.ForeignKey(
        "projects.Project",
        verbose_name=_("project"),
        related_name="children",
        on_delete=models.PROTECT,
    )
    languages_spoken_at_home = models.ManyToManyField(
        Language,
        verbose_name=_("languages spoken at home"),
        related_name="children",
        blank=True,
    )

    objects = ChildQuerySet.as_manager()

    class Meta:
        verbose_name = _("child")
        verbose_name_plural = _("children")
        ordering = ["birthyear", "name"]

    def __str__(self):
        return f"{self.name} ({self.birthyear})"

    @transaction.atomic()
    def delete(self, *args, **kwargs):
        self.enrolments.upcoming().delete()
        return super().delete(*args, **kwargs)

    def can_user_administer(self, user):
        return user.can_administer_project(self.project)

    def get_enrolment_count(self, year: Optional[int] = None, past=False):
        from events.models import Event  # noqa

        if year and past:
            raise ValueError("Cannot use year and past arguments at the same time.")
        now = timezone.now()
        year = year or now.year

        occurrence_filters = Q(time__year=year)
        # For external ticket system events published_at field is used to determine the
        # event's year. That is obviously not a perfect solution, but the best we can
        # do with the current data, and most probably good enough.
        ticket_system_filters = Q(event__published_at__year=year)
        if past:
            occurrence_filters &= Q(time__lt=now)
            ticket_system_filters &= Q(event__published_at__lt=now)
        return (
            self.occurrences.filter(occurrence_filters).count()
            + self.ticket_system_passwords.filter(ticket_system_filters)
            .distinct()
            .count()
        )

    @staticmethod
    def _hash(values: Iterable[str]) -> str:
        return sha256(",".join(values).encode("utf8")).hexdigest()

    @property
    def _salt_birthyear_postal_code_guardian_emails_list(self) -> list[str]:
        """
        List of salt, child's birthyear, postal code and guardians' emails.
        """
        return [
            settings.KUKKUU_HASHID_SALT,
            self.birthyear.isoformat(),
            self.postal_code,
            *sorted(guardian.email for guardian in self.guardians.all()),
        ]

    @property
    def birthyear_postal_code_guardian_emails_hash(self) -> str:
        """
        Salted hash of child's birthyear, postal code and guardians' emails.
        """
        return self._hash(self._salt_birthyear_postal_code_guardian_emails_list)

    @property
    def name_birthyear_postal_code_guardian_emails_hash(self) -> str:
        """
        Salted hash of child's name, birthyear, postal
        code and guardians' emails.
        """
        return self._hash(
            [self.name] + self._salt_birthyear_postal_code_guardian_emails_list
        )


class RelationshipQuerySet(models.QuerySet):
    def user_can_view(self, user):
        return self.filter(
            Q(guardian__user=user) | Q(child__project__in=user.administered_projects)
        ).distinct()


class Relationship(models.Model):
    PARENT = "parent"  # In Finnish: Vanhempi
    OTHER_GUARDIAN = "other_guardian"  # In Finnish: Muu huoltaja
    OTHER_RELATION = "other_relation"  # In Finnish: Muu l√§heinen
    ADVOCATE = "advocate"  # In Finnish: Virallinen puolesta-asioija
    TYPE_CHOICES = (
        (PARENT, _("Parent")),
        (OTHER_GUARDIAN, _("Other guardian")),
        (OTHER_RELATION, _("Other relation")),
        (ADVOCATE, _("Advocate")),
    )

    created_at = models.DateTimeField(verbose_name=_("created at"), auto_now_add=True)
    child = models.ForeignKey(
        Child,
        verbose_name=_("child"),
        related_name="relationships",
        on_delete=models.CASCADE,
    )
    guardian = models.ForeignKey(
        Guardian,
        verbose_name=_("guardian"),
        on_delete=models.CASCADE,
        related_name="relationships",
    )
    type = models.CharField(
        verbose_name=_("type"),
        choices=TYPE_CHOICES,
        max_length=64,
        null=True,
        blank=True,
    )

    objects = RelationshipQuerySet.as_manager()

    class Meta:
        verbose_name = _("relationship")
        verbose_name_plural = _("relationships")
        ordering = [
            "guardian",
            "child",
        ]
