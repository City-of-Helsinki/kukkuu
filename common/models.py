import uuid

from django.db import models
from django.utils.translation import ugettext_lazy as _
from parler.managers import TranslatableQuerySet as ParlerTranslatableQuerySet
from parler.models import TranslatableModel as ParlerTranslatableModel


class TimestampedModel(models.Model):
    created_at = models.DateTimeField(verbose_name=_("created at"), auto_now_add=True)
    updated_at = models.DateTimeField(verbose_name=_("updated_at"), auto_now=True)

    class Meta:
        abstract = True


class UUIDPrimaryKeyModel(models.Model):
    id = models.UUIDField(
        verbose_name=_("UUID"), primary_key=True, default=uuid.uuid4, editable=False
    )

    class Meta:
        abstract = True


class TranslatableQuerySet(ParlerTranslatableQuerySet):
    def create_translatable_object(self, **kwargs):
        translations = kwargs.pop("translations")
        obj = self.create(**kwargs)
        for translation in translations:
            language_code = translation.pop("language_code")
            obj.create_translation(language_code=language_code, **translation)
        return obj


class TranslatableModel(ParlerTranslatableModel):
    objects = TranslatableQuerySet.as_manager()

    class Meta:
        abstract = True
