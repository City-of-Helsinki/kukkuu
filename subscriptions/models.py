from django.db import models
from django.db.models import Q, UniqueConstraint
from django.utils.translation import ugettext_lazy as _

from children.models import Child
from events.models import Occurrence


class FreeSpotNotificationSubscriptionQueryset(models.QuerySet):
    def user_can_view(self, user):
        return self.filter(
            Q(child__guardians__user=user) | Q(child__project__users=user)
        ).distinct()


class FreeSpotNotificationSubscription(models.Model):
    created_at = models.DateTimeField(verbose_name=_("created at"), auto_now_add=True)
    child = models.ForeignKey(
        Child,
        verbose_name=_("child"),
        related_name="free_spot_notification_subscriptions",
        on_delete=models.CASCADE,
    )
    occurrence = models.ForeignKey(
        Occurrence,
        verbose_name=_("occurrence"),
        related_name="free_spot_notification_subscriptions",
        on_delete=models.CASCADE,
    )

    objects = FreeSpotNotificationSubscriptionQueryset.as_manager()

    class Meta:
        verbose_name = _("free spot notification subscription")
        verbose_name_plural = _("free spot notification subscriptions")
        ordering = ("id",)
        constraints = (
            UniqueConstraint(
                fields=["child", "occurrence"], name="unique_free_spot_child_occurrence"
            ),
        )

    def __str__(self):
        return f"{self.child} {self.occurrence} subscription"
