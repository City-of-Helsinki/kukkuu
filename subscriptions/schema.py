import logging

import graphene
from django.core.exceptions import ValidationError
from graphene import relay
from graphene_django import DjangoObjectType

from children.models import Child
from children.schema import ChildNode
from common.utils import (
    get_node_id_from_global_id,
    login_required,
    map_enums_to_values_in_kwargs,
)
from events.models import Occurrence
from events.schema import OccurrenceNode
from hel_django_auditlog_extra.graphene_decorators import auditlog_access
from kukkuu.consts import OCCURRENCE_IS_NOT_FULL_ERROR
from kukkuu.exceptions import (
    AlreadySubscribedError,
    ObjectDoesNotExistError,
    OccurrenceIsNotFullError,
)
from subscriptions.models import FreeSpotNotificationSubscription
from users.schema import GuardianNode
from verification_tokens.decorators import user_from_auth_verification_token

logger = logging.getLogger(__name__)


@auditlog_access
class FreeSpotNotificationSubscriptionNode(DjangoObjectType):
    occurrence = graphene.Field(OccurrenceNode)  # WORKAROUND: See resolve_occurrence

    class Meta:
        model = FreeSpotNotificationSubscription
        interfaces = (relay.Node,)
        fields = ("id", "created_at", "child")
        filter_fields = ("child_id", "occurrence_id")

    @classmethod
    @login_required
    def get_queryset(cls, queryset, info):
        return super().get_queryset(queryset, info).user_can_view(info.context.user)

    def resolve_occurrence(self, info, **kwargs):
        """
        Resolver for the occurrence field, which is a foreign key.

        WORKAROUND: Just putting "occurrence" into Meta.fields breaks
        occurrence field resolving in test_child_free_spot_notifications_query
        with graphene-django 3.2.2, but works with graphene-django 3.1.3.

        Probably caused by the changes done after 3.1.3:
        - "fix: fk resolver permissions leak":
          - https://github.com/graphql-python/graphene-django/pull/1411
        - "fix: foreign key nullable and custom resolver":
          - https://github.com/graphql-python/graphene-django/pull/1446

        Found a similar graphene-django issue #1482:
        - "Vesion 3.1.5 returns null for an object pointed to non PK in referred table":
          - https://github.com/graphql-python/graphene-django/issues/1482
        """
        return self.occurrence


def validate_free_spot_notification_subscription(child, occurrence):
    if child.free_spot_notification_subscriptions.filter(
        occurrence=occurrence
    ).exists():
        raise AlreadySubscribedError(
            "Child already subscribed to free spot notifications of this occurrence"
        )
    try:
        FreeSpotNotificationSubscription(child=child, occurrence=occurrence).clean()
    except ValidationError as e:
        if e.code == OCCURRENCE_IS_NOT_FULL_ERROR:
            raise OccurrenceIsNotFullError(e.message)


def _get_child_and_occurrence(info, **kwargs):
    occurrence_id = get_node_id_from_global_id(
        kwargs["occurrence_id"], "OccurrenceNode"
    )
    child_id = get_node_id_from_global_id(kwargs["child_id"], "ChildNode")
    user = info.context.user
    try:
        child = Child.objects.user_can_update(user).get(pk=child_id)
    except Child.DoesNotExist as e:
        raise ObjectDoesNotExistError(e)

    try:
        occurrence = Occurrence.objects.filter(event__project=child.project).get(
            pk=occurrence_id
        )
    except Occurrence.DoesNotExist as e:
        raise ObjectDoesNotExistError(e)

    return child, occurrence


class SubscribeToFreeSpotNotificationMutation(graphene.relay.ClientIDMutation):
    class Input:
        occurrence_id = graphene.GlobalID()
        child_id = graphene.GlobalID()

    occurrence = graphene.Field(OccurrenceNode)
    child = graphene.Field(ChildNode)

    @classmethod
    @login_required
    @map_enums_to_values_in_kwargs
    def mutate_and_get_payload(cls, root, info, **kwargs):
        user = info.context.user
        child, occurrence = _get_child_and_occurrence(info, **kwargs)

        validate_free_spot_notification_subscription(child, occurrence)

        FreeSpotNotificationSubscription.objects.create(
            child=child, occurrence=occurrence
        )

        logger.info(
            f"User {user.uuid} subscribed child {child.pk} to occurrence "
            f"{occurrence} free spot notification."
        )

        return SubscribeToFreeSpotNotificationMutation(
            child=child, occurrence=occurrence
        )


class UnsubscribeFromFreeSpotNotificationMutation(graphene.relay.ClientIDMutation):
    class Input:
        occurrence_id = graphene.GlobalID()
        child_id = graphene.GlobalID()

    occurrence = graphene.Field(OccurrenceNode)
    child = graphene.Field(ChildNode)

    @classmethod
    @login_required
    @map_enums_to_values_in_kwargs
    def mutate_and_get_payload(cls, root, info, **kwargs):
        user = info.context.user
        child, occurrence = _get_child_and_occurrence(info, **kwargs)

        try:
            subscription = FreeSpotNotificationSubscription.objects.get(
                child=child, occurrence=occurrence
            )
        except FreeSpotNotificationSubscription.DoesNotExist as e:
            raise ObjectDoesNotExistError(e)

        subscription.delete()

        logger.info(
            f"User {user.uuid} unsubscribed child {child.pk} from occurrence "
            f"{occurrence} free spot notification."
        )

        return UnsubscribeFromFreeSpotNotificationMutation(
            child=child, occurrence=occurrence
        )


class UnsubscribeFromAllNotificationsMutation(graphene.relay.ClientIDMutation):
    """
    Unsubscribe user from all the notifications.

    NOTE: This mutation deletes the user's FreeSpotNotifications,
    which are linked to a Child and Occurrence instances.
    **It should be noted that the current model architecture allows
    that a child can have multiple guardians, so unsubscribe can delete
    some notifications from other users as well. However, the UI apps
    has never allowed more than 1 guardian for a child.**
    """

    class Input:
        auth_token = graphene.String(
            description="Auth token can be used to authorize the action "
            "without logging in as a user."
        )

    guardian = graphene.Field(GuardianNode)
    unsubscribed = graphene.Boolean()

    @classmethod
    # When the login_required raises a PermissionDenied exception,
    # use the auth_token from the input variables
    # to populate the context.user with the token related user
    # and then try again.
    @user_from_auth_verification_token(
        get_token=lambda variables: variables.get("auth_token", None),
        use_only_when_first_denied=True,
    )
    @login_required
    @map_enums_to_values_in_kwargs
    def mutate_and_get_payload(cls, root, info, **kwargs):
        user = info.context.user
        user.unsubscribe_all_notification_subscriptions()
        logger.info(f"User {user.uuid} unsubscribed from all notifications.")
        return UnsubscribeFromAllNotificationsMutation(
            guardian=user.guardian, unsubscribed=True
        )


class Mutation:
    subscribe_to_free_spot_notification = (
        SubscribeToFreeSpotNotificationMutation.Field()
    )
    unsubscribe_from_free_spot_notification = (
        UnsubscribeFromFreeSpotNotificationMutation.Field()
    )
    unsubscribe_from_all_notifications = UnsubscribeFromAllNotificationsMutation.Field()
