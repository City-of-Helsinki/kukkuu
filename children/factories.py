import factory

from children.models import Child, Relationship
from projects.models import Project
from users.factories import GuardianFactory


class ChildFactory(factory.django.DjangoModelFactory):
    name = factory.Faker("name")
    birthyear = factory.Faker(
        "random_int", min=2018, max=2023
    )  # NOTE: Many of the tests are freezed to 2020
    postal_code = factory.Faker("postcode", locale="fi_FI")
    project = factory.LazyFunction(lambda: Project.objects.get(year=2020))
    notes = factory.Faker(
        "random_element",
        elements=[
            "",
            "Test notes",
            "Alternative test notes",
            "Longer test notes\nwith multiple lines.",
        ],
    )

    class Meta:
        model = Child


class RelationshipFactory(factory.django.DjangoModelFactory):
    child = factory.SubFactory(ChildFactory)
    guardian = factory.SubFactory(GuardianFactory)
    type = factory.Faker(
        "random_element", elements=[t[0] for t in Relationship.TYPE_CHOICES]
    )

    class Meta:
        model = Relationship


class ChildWithGuardianFactory(ChildFactory):
    relationship = factory.RelatedFactory(RelationshipFactory, "child")
    project = factory.LazyFunction(lambda: Project.objects.get(year=2020))


class ChildWithTwoGuardiansFactory(ChildWithGuardianFactory):
    relationship2 = factory.RelatedFactory(RelationshipFactory, "child")
