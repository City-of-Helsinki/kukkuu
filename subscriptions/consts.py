class NotificationType:
    FREE_SPOT = "free_spot"
