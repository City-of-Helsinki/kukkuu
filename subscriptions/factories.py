import factory

from children.factories import ChildFactory
from events.factories import OccurrenceFactory
from subscriptions.models import FreeSpotNotificationSubscription


class FreeSpotNotificationSubscriptionFactory(factory.django.DjangoModelFactory):
    occurrence = factory.SubFactory(OccurrenceFactory)
    child = factory.SubFactory(ChildFactory)

    class Meta:
        model = FreeSpotNotificationSubscription
        skip_postgeneration_save = True  # Not needed after factory v4.0.0
