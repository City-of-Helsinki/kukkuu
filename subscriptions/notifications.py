from django.utils import timezone
from django.utils.translation import gettext_lazy as _
from django_ilmoitin.dummy_context import dummy_context
from django_ilmoitin.registry import notifications

from children.factories import ChildWithGuardianFactory
from events.factories import EventFactory, OccurrenceFactory
from events.utils import (
    get_event_ui_url,
    get_occurrence_enrol_ui_url,
    get_occurrence_ui_url,
)
from projects.factories import ProjectFactory
from subscriptions.consts import NotificationType
from subscriptions.factories import FreeSpotNotificationSubscriptionFactory
from users.factories import GuardianFactory
from venues.factories import VenueFactory

notifications.register(NotificationType.FREE_SPOT, _("free spot"))

project = ProjectFactory.build(year=2020)
event = EventFactory.build(project=project)
venue = VenueFactory.build(project=project)
guardian = GuardianFactory.build()
child = ChildWithGuardianFactory.build(relationship__guardian=guardian, project=project)
occurrence = OccurrenceFactory.build(event=event, venue=venue)
subscription = FreeSpotNotificationSubscriptionFactory.build(
    child=child, occurrence=occurrence
)
unsubscribe_url = "https://kukkuu-ui-domain/fi/profile/subscriptions?authToken=abc123"
dummy_context.update(
    {
        NotificationType.FREE_SPOT: {
            "guardian": guardian,
            "event": event,
            "child": child,
            "event_url": get_event_ui_url(event, child, guardian.language),
            "occurrence": occurrence,
            "occurrence_url": get_occurrence_ui_url(
                occurrence, child, guardian.language
            ),
            "occurrence_enrol_url": get_occurrence_enrol_ui_url(
                occurrence, child, guardian.language
            ),
            "subscription": subscription,
            "localtime": timezone.template_localtime,
            "unsubscribe_url": unsubscribe_url,
        },
    }
)
