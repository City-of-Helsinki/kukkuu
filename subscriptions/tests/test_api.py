from datetime import timedelta

import pytest
from django.utils.timezone import now

from children.factories import ChildWithGuardianFactory
from common.tests.utils import assert_match_error_code
from common.utils import get_global_id
from events.factories import EventFactory, OccurrenceFactory
from kukkuu.consts import (
    ALREADY_SUBSCRIBED_ERROR,
    OBJECT_DOES_NOT_EXIST_ERROR,
    OCCURRENCE_IS_NOT_FULL_ERROR,
    PERMISSION_DENIED_ERROR,
)
from subscriptions.factories import FreeSpotNotificationSubscriptionFactory
from subscriptions.models import FreeSpotNotificationSubscription
from users.factories import GuardianFactory


@pytest.fixture(autouse=True)
def autouse_db(db):
    pass


CHILD_FREE_SPOT_NOTIFICATION_SUBSCRIPTIONS_QUERY = """
query ChildFreeSpotNotificationSubscriptionsQuery($id: ID!) {
  child(id: $id) {
    freeSpotNotificationSubscriptions {
      edges {
        node {
          createdAt
          id
          child {
            name
          }
          occurrence {
            time
          }
        }
      }
    }
  }
}
"""

OCCURRENCES_HAS_CHILD_FREE_SPOT_NOTIFICATION_SUBSCRIPTION_QUERY = """
query OccurrencesHasChildFreeSpotNotificationSubscription($childId: ID!) {
  child(id: $childId) {
    availableEvents {
      edges {
        node {
            occurrences {
              edges {
                node {
                  time
                  childHasFreeSpotNotificationSubscription(childId: $childId)
                }
              }
            }
        }
      }
    }
  }
}
"""


@pytest.fixture
def guardian_child(guardian_api_client):
    return ChildWithGuardianFactory(
        name="Subscriber",
        relationship__guardian__user=guardian_api_client.user,
    )


def test_child_free_spot_notifications_query(
    snapshot, guardian_api_client, guardian_child
):
    FreeSpotNotificationSubscriptionFactory(id=1, child=guardian_child)
    executed = guardian_api_client.execute(
        CHILD_FREE_SPOT_NOTIFICATION_SUBSCRIPTIONS_QUERY,
        variables={"id": get_global_id(guardian_child)},
    )

    snapshot.assert_match(executed)


def test_occurrences_has_child_free_spot_notification_query(
    snapshot, guardian_api_client, guardian_child
):
    event = EventFactory(published_at=now())
    FreeSpotNotificationSubscriptionFactory(
        child=guardian_child,
        occurrence__event=event,
        occurrence__time=now() + timedelta(days=14),
    )
    FreeSpotNotificationSubscriptionFactory(
        occurrence__event=event, occurrence__time=now() + timedelta(days=15)
    )

    executed = guardian_api_client.execute(
        OCCURRENCES_HAS_CHILD_FREE_SPOT_NOTIFICATION_SUBSCRIPTION_QUERY,
        variables={"childId": get_global_id(guardian_child)},
    )

    snapshot.assert_match(executed)


SUBSCRIBE_TO_FREE_SPOT_NOTIFICATION_MUTATION = """
mutation SubscribeToFreeSpotNotification($input: SubscribeToFreeSpotNotificationMutationInput!) {
  subscribeToFreeSpotNotification(input: $input) {
    child {
      name
    }
    occurrence {
      time
    }
  }
}

"""  # noqa


def test_subscribe_to_free_spot_notification(
    snapshot, guardian_api_client, guardian_child
):
    occurrence = OccurrenceFactory(
        event__published_at=now(),
        time=now() + timedelta(days=14),
        event__capacity_per_occurrence=0,
    )

    executed = guardian_api_client.execute(
        SUBSCRIBE_TO_FREE_SPOT_NOTIFICATION_MUTATION,
        variables={
            "input": {
                "occurrenceId": get_global_id(occurrence),
                "childId": get_global_id(guardian_child),
            }
        },
    )

    snapshot.assert_match(executed)
    assert FreeSpotNotificationSubscription.objects.get(
        child=guardian_child, occurrence=occurrence
    )

    executed = guardian_api_client.execute(
        SUBSCRIBE_TO_FREE_SPOT_NOTIFICATION_MUTATION,
        variables={
            "input": {
                "occurrenceId": get_global_id(occurrence),
                "childId": get_global_id(guardian_child),
            }
        },
    )

    assert_match_error_code(executed, ALREADY_SUBSCRIBED_ERROR)


def test_cannot_subscribe_to_free_spot_notification_with_someone_elses_child(
    guardian_api_client,
):
    occurrence = OccurrenceFactory(
        event__published_at=now(), time=now() + timedelta(days=14)
    )
    child = ChildWithGuardianFactory(name="Subscriber")

    executed = guardian_api_client.execute(
        SUBSCRIBE_TO_FREE_SPOT_NOTIFICATION_MUTATION,
        variables={
            "input": {
                "occurrenceId": get_global_id(occurrence),
                "childId": get_global_id(child),
            }
        },
    )

    assert_match_error_code(executed, OBJECT_DOES_NOT_EXIST_ERROR)


def test_cannot_subscribe_to_free_spot_notification_when_occurrence_not_full(
    guardian_api_client, guardian_child
):
    occurrence = OccurrenceFactory(
        event__published_at=now(),
        time=now() + timedelta(days=14),
        capacity_override=1,
    )

    executed = guardian_api_client.execute(
        SUBSCRIBE_TO_FREE_SPOT_NOTIFICATION_MUTATION,
        variables={
            "input": {
                "occurrenceId": get_global_id(occurrence),
                "childId": get_global_id(guardian_child),
            }
        },
    )

    assert_match_error_code(executed, OCCURRENCE_IS_NOT_FULL_ERROR)


UNSUBSCRIBE_FROM_FREE_SPOT_NOTIFICATION_MUTATION = """
mutation UnsubscribeFromFreeSpotNotification($input: UnsubscribeFromFreeSpotNotificationMutationInput!) {
  unsubscribeFromFreeSpotNotification(input: $input) {
    child {
      name
    }
    occurrence {
      time
    }
  }
}

"""  # noqa


def test_unsubscribe_from_free_spot_notification(
    snapshot, guardian_api_client, guardian_child
):
    subscription = FreeSpotNotificationSubscriptionFactory(child=guardian_child)

    executed = guardian_api_client.execute(
        UNSUBSCRIBE_FROM_FREE_SPOT_NOTIFICATION_MUTATION,
        variables={
            "input": {
                "occurrenceId": get_global_id(subscription.occurrence),
                "childId": get_global_id(guardian_child),
            }
        },
    )

    snapshot.assert_match(executed)
    assert not FreeSpotNotificationSubscription.objects.filter(
        child=guardian_child, occurrence=subscription.occurrence
    ).exists()


def test_cannot_unsubscribe_from_free_spot_notification_with_someone_elses_child(
    guardian_api_client,
):
    child = ChildWithGuardianFactory(name="Subscriber")
    subscription = FreeSpotNotificationSubscriptionFactory(child=child)

    executed = guardian_api_client.execute(
        UNSUBSCRIBE_FROM_FREE_SPOT_NOTIFICATION_MUTATION,
        variables={
            "input": {
                "occurrenceId": get_global_id(subscription.occurrence),
                "childId": get_global_id(child),
            }
        },
    )

    assert_match_error_code(executed, OBJECT_DOES_NOT_EXIST_ERROR)


UNSUBSCRIBE_FROM_ALL_NOTIFICATIONS_MUTATION = """
mutation UnsubscribeFromAllNotifications(
    $input: UnsubscribeFromAllNotificationsMutationInput!
) {
  unsubscribeFromAllNotifications(input: $input) {
    guardian {
        user {
            username
        }
    }
    unsubscribed
  }
}
"""


def test_unsubscribe_from_all_notifications(
    snapshot, guardian_api_client, guardian_child
):
    guardian = guardian_child.guardians.first()
    user = guardian.user
    FreeSpotNotificationSubscriptionFactory(child=guardian_child)
    executed = guardian_api_client.execute(
        UNSUBSCRIBE_FROM_ALL_NOTIFICATIONS_MUTATION,
        variables={"input": {}},
    )
    assert (
        executed["data"]["unsubscribeFromAllNotifications"]["guardian"]["user"][
            "username"
        ]
        == user.username
    )
    snapshot.assert_match(executed)
    assert not FreeSpotNotificationSubscription.objects.filter(
        child=guardian_child
    ).exists()


def test_unsubscribe_from_all_notifications_using_auth_verification_token(
    snapshot, api_client, guardian_child
):
    guardian = guardian_child.guardians.first()
    user = guardian.user
    FreeSpotNotificationSubscriptionFactory(child=guardian_child)
    auth_verification_token = user.create_subscriptions_management_auth_token()
    executed = api_client.execute(
        UNSUBSCRIBE_FROM_ALL_NOTIFICATIONS_MUTATION,
        variables={
            "input": {
                "authToken": auth_verification_token.key,
            }
        },
    )
    assert (
        executed["data"]["unsubscribeFromAllNotifications"]["guardian"]["user"][
            "username"
        ]
        == user.username
    )
    snapshot.assert_match(executed)
    assert not FreeSpotNotificationSubscription.objects.filter(
        child=guardian_child
    ).exists()


def test_unsubscribe_from_all_notifications_using_invalid_auth_verification_token(
    api_client, guardian_child
):
    guardian = guardian_child.guardians.first()
    user = guardian.user
    FreeSpotNotificationSubscriptionFactory(child=guardian_child)
    user.create_subscriptions_management_auth_token()
    executed = api_client.execute(
        UNSUBSCRIBE_FROM_ALL_NOTIFICATIONS_MUTATION,
        variables={
            "input": {
                "authToken": "invalid token",
            }
        },
    )
    assert_match_error_code(executed, PERMISSION_DENIED_ERROR)
    assert FreeSpotNotificationSubscription.objects.filter(
        child=guardian_child
    ).exists()


def test_unsubscribe_from_all_notifications_using_auth_verification_token_as_logged_in(
    snapshot, guardian_api_client, guardian_child
):
    logged_in_guardian = guardian_api_client.user.guardian
    child_guardian = guardian_child.guardians.first()

    assert logged_in_guardian.id == child_guardian.id

    user = child_guardian.user
    FreeSpotNotificationSubscriptionFactory(child=guardian_child)

    # Token is created for another user
    auth_verification_token = user.create_subscriptions_management_auth_token()
    executed = guardian_api_client.execute(
        UNSUBSCRIBE_FROM_ALL_NOTIFICATIONS_MUTATION,
        variables={
            "input": {
                "authToken": auth_verification_token.key,
            }
        },
    )
    assert (
        executed["data"]["unsubscribeFromAllNotifications"]["guardian"]["user"][
            "username"
        ]
        == user.username
    )
    snapshot.assert_match(executed)
    assert not FreeSpotNotificationSubscription.objects.filter(
        child=guardian_child
    ).exists()


def test_unsubscribe_from_all_notifications_when_logged_in_user_not_auth_token_user(
    snapshot, user_api_client, guardian_child
):
    logged_in_guardian = GuardianFactory(user=user_api_client.user)
    child_guardian = guardian_child.guardians.first()

    assert logged_in_guardian != child_guardian

    FreeSpotNotificationSubscriptionFactory(child=guardian_child)

    # Token is created for another user
    auth_verification_token = (
        child_guardian.user.create_subscriptions_management_auth_token()
    )
    executed = user_api_client.execute(
        UNSUBSCRIBE_FROM_ALL_NOTIFICATIONS_MUTATION,
        variables={
            "input": {
                "authToken": auth_verification_token.key,
            }
        },
    )
    assert (
        executed["data"]["unsubscribeFromAllNotifications"]["guardian"]["user"][
            "username"
        ]
        == logged_in_guardian.user.username
    )
    snapshot.assert_match(executed)
    assert FreeSpotNotificationSubscription.objects.filter(
        child=guardian_child
    ).exists()
