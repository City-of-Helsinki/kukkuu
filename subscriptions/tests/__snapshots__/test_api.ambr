# serializer version: 1
# name: test_child_free_spot_notifications_query
  dict({
    'data': dict({
      'child': dict({
        'freeSpotNotificationSubscriptions': dict({
          'edges': list([
            dict({
              'node': dict({
                'child': dict({
                  'name': 'Subscriber',
                }),
                'createdAt': '2020-12-12T00:00:00+00:00',
                'id': 'RnJlZVNwb3ROb3RpZmljYXRpb25TdWJzY3JpcHRpb25Ob2RlOjE=',
                'occurrence': dict({
                  'time': '2012-11-14T05:07:37.553889+00:00',
                }),
              }),
            }),
          ]),
        }),
      }),
    }),
  })
# ---
# name: test_occurrences_has_child_free_spot_notification_query
  dict({
    'data': dict({
      'child': dict({
        'availableEvents': dict({
          'edges': list([
            dict({
              'node': dict({
                'occurrences': dict({
                  'edges': list([
                    dict({
                      'node': dict({
                        'childHasFreeSpotNotificationSubscription': True,
                        'time': '2020-12-26T00:00:00+00:00',
                      }),
                    }),
                    dict({
                      'node': dict({
                        'childHasFreeSpotNotificationSubscription': False,
                        'time': '2020-12-27T00:00:00+00:00',
                      }),
                    }),
                  ]),
                }),
              }),
            }),
          ]),
        }),
      }),
    }),
  })
# ---
# name: test_subscribe_to_free_spot_notification
  dict({
    'data': dict({
      'subscribeToFreeSpotNotification': dict({
        'child': dict({
          'name': 'Subscriber',
        }),
        'occurrence': dict({
          'time': '2020-12-26T00:00:00+00:00',
        }),
      }),
    }),
  })
# ---
# name: test_unsubscribe_from_all_notifications
  dict({
    'data': dict({
      'unsubscribeFromAllNotifications': dict({
        'guardian': dict({
          'user': dict({
            'username': 'jeffersonkimberly',
          }),
        }),
        'unsubscribed': True,
      }),
    }),
  })
# ---
# name: test_unsubscribe_from_all_notifications_using_auth_verification_token
  dict({
    'data': dict({
      'unsubscribeFromAllNotifications': dict({
        'guardian': dict({
          'user': dict({
            'username': 'jeffersonkimberly',
          }),
        }),
        'unsubscribed': True,
      }),
    }),
  })
# ---
# name: test_unsubscribe_from_all_notifications_using_auth_verification_token_as_logged_in
  dict({
    'data': dict({
      'unsubscribeFromAllNotifications': dict({
        'guardian': dict({
          'user': dict({
            'username': 'jeffersonkimberly',
          }),
        }),
        'unsubscribed': True,
      }),
    }),
  })
# ---
# name: test_unsubscribe_from_all_notifications_when_logged_in_user_not_auth_token_user
  dict({
    'data': dict({
      'unsubscribeFromAllNotifications': dict({
        'guardian': dict({
          'user': dict({
            'username': 'jeffersonkimberly',
          }),
        }),
        'unsubscribed': True,
      }),
    }),
  })
# ---
# name: test_unsubscribe_from_free_spot_notification
  dict({
    'data': dict({
      'unsubscribeFromFreeSpotNotification': dict({
        'child': dict({
          'name': 'Subscriber',
        }),
        'occurrence': dict({
          'time': '2012-11-14T05:07:37.553889+00:00',
        }),
      }),
    }),
  })
# ---
